Use link tokens instead of public key for linking/relinking accounts

jdormit opened the pull request:
This is an attempt at resolving #1  - using link tokens instead of public keys to access the Plaid API. I basically followed the template you suggested - updating the `plaid-go` dependency to the latest version and changing the link and relink flows to use the new link token methods. I don't have much Go experience, so any feedback is welcome.

I had to break out a totally separate `relink` method in the linker, since in the new flow relinking doesn't return anything on success - the access token the app previously had just stays valid. This made the linker more verbose than it was before, but I couldn't think of a better way to represent the new logic.

Let me know what you think!

jdormit committed:
Use link tokens instead of public key for linking/relinking accounts

jdormit commented on main.go:
```diff
@@ -102,13 +101,18 @@ func main() {
-				tokenPair, err = linker.Relink(itemOrAlias, port)
+				err = linker.Relink(itemOrAlias, port)
+				log.Println("Institution relinked!")
+				return
```
Just returning here, as in the new relink flow no new access token is returned.

jdormit commented on main.go:
```diff
@@ -334,7 +338,7 @@ func main() {
 					IncludeStatus:           withStatusFlag,
 				}
-				resp, err := client.GetInstitutionByIDWithOptions(instID, opts)
+				resp, err := client.GetInstitutionByIDWithOptions(instID, []string {"US"}, opts)
```
New `plaid-go` API required this additional argument - I figure just hard-coding `"US"` is fine?

landakram replied:
Could we used [`go-locale`](https://github.com/Xuanwo/go-locale) to detect the locale and infer language and country? It would be nice to keep things country-agnostic and I don't mind bringing in that dependency

jdormit commented on pkg/plaid_cli/linker.go:
```diff
@@ -5,44 +5,75 @@ import (
+func (l *Linker) Relink(itemID string, port string) error {
 	token := l.Data.Tokens[itemID]
-	res, err := l.Client.CreatePublicToken(token)
+	hostname, err := os.Hostname()
```
The LinkToken API requires a unique user id. I figured we can just use the hostname here, but maybe we want this to be something the user can configure? What do you think?

landakram replied:
That seems fine to me for now. Ideally we'd store it in the `Data` object so it's configurable, but that change can be made  later if/when it becomes an issue.

jdormit commented on pkg/plaid_cli/linker.go:
```diff
@@ -5,44 +5,75 @@ import (
+			ClientUserID: hostname,
+		},
+		ClientName:   "plaid-cli",
+		Products:     []string{"transactions"},
```
Does `plaid-cli` require access to any of the other Plaid products? Options are `transactions, auth, identity, assets, investments, liabilities, payment_initiation`.

landakram replied:
`"transactions"` should be fine for now. I just got access to Investments, but we can change later 😛

jdormit commented on pkg/plaid_cli/linker.go:
```diff
@@ -125,47 +169,31 @@ func handleLink(linker *Linker) func(w http.ResponseWriter, r *http.Request) {
 
 type LinkTmplData struct {
-	PublicKey   string
-	Environment string
```
I couldn't find anywhere in the new link flow Plaid docs that mentioned specifying which environment you are in. Maybe this is now implied by the client credentials?

landakram replied:
Seems right to me 👍

jdormit committed:
Format

jdormit commented on pkg/plaid_cli/linker.go:
```diff
@@ -179,19 +207,19 @@ var linkTemplate string = `<html>
   <head>
     <style>
     .alert-success {
-        font-size: 1.2em;
```
😑 I guess Emacs decided to make some whitespace changes. Happy to revert these if that's not welcome.

landakram replied:
No worries!

jdormit commented on pkg/plaid_cli/linker.go:
```diff
@@ -289,56 +293,28 @@ var relinkTemplate string = `<html>
-           // }
-         }
+	 token: '{{ .LinkToken }}',
+	 onSuccess: (public_token, metadata) => {
```
For some reason Plaid requires this `onSuccess` method to be defined, but doesn't seem to actually call it for a relink. Go figure...

landakram commented on pkg/plaid_cli/linker.go:
```diff
@@ -5,44 +5,75 @@ import (
+			ClientUserID: hostname,
+		},
+		ClientName:   "plaid-cli",
+		CountryCodes: []string{"US"},
```
See my comment above regarding go-locale.

landakram commented on pkg/plaid_cli/linker.go:
```diff
@@ -5,44 +5,75 @@ import (
+		},
+		ClientName:   "plaid-cli",
+		Products:     []string{"transactions"},
+		CountryCodes: []string{"US"},
```
See my comment above regarding go-locale.

landakram commented on pkg/plaid_cli/linker.go:
```diff
@@ -5,44 +5,75 @@ import (
 	}
-
-	return l.link(port, handleRelink(l, res.PublicToken))
+	client, err := plaid.NewClient(l.ClientOpts)
```
Is there a reason to use `plaid.NewClient` rather than the `l.Client` object that gets passed into the constructor?

I can't quite remember why I saved `l.ClientOpts`, but I'm pretty sure it was to deal with the `"production"` vs. `"development"` environment strings below. Since those are no longer necessary and if using `l.Client` suffices, let's remove `l.ClientOpts` entirely and use `l.Client` everywhere.

landakram reviewed (commented):
This is looking really good. Thanks for submitting the PR 🙂 

I had some small suggestions about detecting locale rather than hardcoding and using the `Linker.Client` object rather than constructing the plaid client with `plaid.NewClient`. If the locale stuff isn't straightforward, then we can merge after addressing the `plaid.NewClient` stuff.

landakram committed:
Let language and countries be configurable

- Language and countries are configurable, but autodetected by default
- Remove Linker.ClientOpts

landakram commented:
@jdormit, I went ahead and made the revisions. Thank you so much for your contributions!

jdormit commented:
Thanks @landakram! Sorry I didn't get to it sooner - my schedule got a bit crazy.

landakram commented:
@jdormit, no worries at all. Do let me know if you end up working on a ledger-cli integration. Would love to help out.

jdormit commented:
I will definitely let you know! It'll likely be a few weeks before I can get back to it.
